docs/smart_bits.md

```diff
@@ -0,0 +1,71 @@
+
+### Bit-width
+The rule of thumb for context-determined expressions is that all it's operands get extended to the **longest** operand appearing in the complete expression (including the LHS of the assignment). On the other hand, the scope for self-determined expressions is only the partial expression itself. Note that since a context-determined expression can appear within a self-determined expression, each self-determined expression must still be procssed by the inference step, except that the context is limited to just the expression itself, and in particular, the LHS of the assignment is excluded.
+
+The bit-width of any partial expression's result is determined by the specific operator (see table below). For example the result of addition has bit-width equal to the maximum of all of it's operand bit-widths, where as the result of comparison always has bit-width equal to 1.
+
+### Signedness
+The signedness of operands can both determine the semantics of operators, as well as how operands are extended. For example, signed-comparison has different semantics than unsigned-comparison, and operands can either be sign-extended or zero-extended (during bit-width inference).
+
+The rule of thumb for both self-determined and context-determined expressions is that all it's operands must be signed for the operator to be interpreted as signed (and therefore produce a signed output). There are the following exceptions:
+* The result of concat is always unsigned
+* The result of comparisons is always unsigned
+
+Additionally, there are two conversion functions (`signed()`, `unsigned()`) which coerce results of partial expressions to be either signed or unsigned, respectively. Note that these functions **do not** affect the signedness of any operands of the partial expression being coerced, but do affect the signedness of any other expressions in which the partial expression appears. For example, in `x + signed(y + z)` the signedness of `y` and `z` is not affected, but `(y + z)` is interpreted as a signed value in the expression `x + (y + z)`.
+
+### Summary of bit-width and signedness rules
+| Expression             | Operands           | Output bit-width | Output signedness       |
+|------------------------|--------------------|------------------|-------------------------|
+| Binary arith.,logical  | Context-determined | max(operands)  | all(signed(operands)) |
+| Unary logical          | Context-determined | max(operands)  | all(signed(operands)) |
+| Shift                  | Shiftee context-determined, shifter self-determined, | Shiftee bit-width | all(signed(operands)) |
+| Comparison             | Context-determined | 1                | unsigned                |
+| Reduction              | Self-determined    | 1                | all(signed(operands)) |
+| Concat                 | Self-determined    | sum(operands)  | unsigned |
+
```

Comment:
What is the signedness of `int` constants?

Reply:
Unfortunately, `int` constants aren't supported operands of smart operators (presumably would be a blocker for actual usage...). But when it is added (soon), they will likely be unsigned. From the Verilog spec:

> * Expression type depends only on the operands. It does not depend on the left-hand side (if any).
> * Decimal numbers are signed.
> * Based_numbers  are  unsigned,  except  where  the  s  notation  is  used  in  the  base  specifier  (as  in"4'sd12").

Reply:
Gotcha.  I would advocate making them signed.  That way there treated as unsigned in an unsigned context and signed in a signed one.  Probably my most commonly used int constant, other than 0, is -1 (for building masks) and I would definitely want it be sign extended.

Reply:
That's a good point. Since we do not have the luxury of having base specifiers for integral constants e.g. `4'sd12` vs `4'd12`, it makes sense to make them signed by default. I will do that when the feature is added.